### ISSUE TITLE
TensorFlow: "Preparing working directory" timed out on mac (also slow on Windows)

### ISSUE THREAD

**Issue by meteorcloudy** [bazelbuild/continuous-integration#602]
https://buildkite.com/bazel/pcloudy-test/builds/38#fed79d19-106a-44c9-b701-7584c6a2929b

It took 25 mins to finish on Windows and timed out on mac after three tries.

**Comment by meteorcloudy** [comment]
Looks like it's now timing out on Windows as well:
https://buildkite.com/bazel/bazel-at-head-plus-disabled/builds/141#e6af873c-ba31-4402-b4be-1600ee21211c

**Comment by philwo** [comment]
Looks like this: https://stackoverflow.com/questions/38618885/error-rpc-failed-curl-transfer-closed-with-outstanding-read-data-remaining

**Comment by philwo** [comment]
Looks like Buildkite is about to add support for shallow clones, which might help here: https://github.com/buildkite/agent/pull/957

We could also look into using SSH instead of HTTP for the connection to GitHub, but it might be annoying to distribute the secret key to all agents...

**Comment by meteorcloudy** [comment]
Can we do the shallow clone in our script?
https://github.com/bazelbuild/continuous-integration/blob/8189703667291f2ec161744747caaeeaff4b9298/buildkite/bazelci.py#L908

**Comment by philwo** [comment]
We can try (we have to implement it there anyway, I think), but that would only help with downstream projects, wouldn't it?

**Comment by meteorcloudy** [comment]
Oh, you're right. 
Is updating the git reference going to help?
`--reference c:\buildkite\bazelbuild `

**Comment by philwo** [comment]
We can try to update the reference bundle on GCS, but I'm not sure if it will help...

**Comment by philwo** [comment]
I'll update it now.

**Comment by meteorcloudy** [comment]
Worth a try, thanks!

**Comment by philwo** [comment]
I've updated it, but we have to do some work to actually use it:

- [ ] Create new Windows images and roll them out.
- [ ] Reboot Docker VMs so that they pick up the new bundle.
- [ ] Manually roll out to macOS machines.

**Comment by philwo** [comment]
All done - the new Gitbundle should be used by all jobs that run from now on!

**Comment by meteorcloudy** [comment]
Nice, it's working! https://buildkite.com/bazel/tensorflow/builds/2556#477b266e-4e22-4272-8410-5dba81eda90c

**Comment by lox** [comment]
Just an FYI, we added support for reference clones as an experiment in the agent recently: https://github.com/buildkite/agent/pull/936.

It works great in the couple of weeks of usage it's gotten.

I'm merging `git-fetch-flags` which will allow shallow clones, which is the other piece of the puzzle for handling huge git repos.

**Comment by meteorcloudy** [comment]
This is happening for downstream again:
https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/950#77dce900-b18e-46d1-a2a7-6e42b11f0be2
@philwo

**Comment by philwo** [comment]
@meteorcloudy Yes, we don't have a known fix for this. :( I don't know what's up with the TensorFlow repository that it always runs into this...

**Comment by meteorcloudy** [comment]
Is shallow clone going to work?
```
$ git clone http://github.com/large-repository --depth 1
$ cd large-repository
$ git fetch --unshallow
```

**Comment by lox** [comment]
You beat me to that suggestion by 10 seconds @meteorcloudy :)

**Comment by philwo** [comment]
I debugged a bit more. We're using an "all-in-one" Git reference bundle on our CI, that I create by fetching remotes of every Git repo that we test on CI into one local Git repo. Apparently, that causes this problem when cloning Tensorflow, as it works fine when just removing the `--reference` flag. It's very fast then.

When I use that clone as a reference for another Tensorflow clone, it's also very fast. So the issue is only caused by using the huge "all-in-one" repo as a reference.

Luckily there's a replacement ready - we could just use Buildkite's native Git mirror support as proposed earlier by @lox :) I'll give this a try.
